File: tests/test-minecraft/src/main/java/pize/tests/minecraft/game/gui/text/TranslationComponent.java
package pize.tests.minecraft.game.gui.text;

import pize.tests.minecraft.game.Session;
import pize.tests.minecraft.game.gui.text.formatting.Style;
import pize.tests.minecraft.game.lang.ClientLanguage;

public class TranslationComponent extends Component{

    private final Component parent;

    private final String key;
    private final Component[] args;
    private final int[] argsIndices;

    private ClientLanguage decomposedWith;

    protected TranslationComponent(Component parent, String key, Component... args){
        this.parent = parent;

        this.key = key;
        this.args = args;
        argsIndices = new int[args.length];
    }


    private void decompose(){
        super.components.clear();
        super.color.set(parent.color);
        System.arraycopy(parent.style, 0, super.style, 0, 6);

        final String[] parts = decomposedWith.getOrDefault(key).split("%s");

        int argIndex = 0;
        for(String part: parts){
            super.formattedText(part);
            if(argIndex < args.length){
                argsIndices[argIndex] = super.components.size() - 1;
                component(args[argIndex]);
                argIndex++;
            }
        }
    }

    public void update(Session session){
        ClientLanguage currentLanguage = session.getLanguageManager().getLanguage();
        if(decomposedWith != currentLanguage){
            decomposedWith = currentLanguage;
            decompose();
        }
    }


    public Component getArg(int index){
        return args[argsIndices[index]];
    }
    
    public Component[] getArgs(){
        return args;
    }

    public String getKey(){
        return key;
    }

    public Style getStyle(){
        return Style.DEFAULT;
    }

}

File: tests/test-minecraft/src/main/kotlin/pize/tests/minecraft/game/gui/text/TranslationComponent.kt
package pize.tests.minecraft.game.gui.text

import pize.tests.minecraft.game.Session
import pize.tests.minecraft.game.gui.text.formatting.Style
import pize.tests.minecraft.game.lang.ClientLanguage

class TranslationComponent(private val parent: Component, val key: String?, vararg args: Component) : Component() {
    val args: Array<Component>
    private val argsIndices: IntArray
    private var decomposedWith: ClientLanguage? = null

    init {
        this.args = args
        argsIndices = IntArray(args.size)
    }

    private fun decompose() {
        super.components.clear()
        super.color.set(parent.color)
        System.arraycopy(parent.style, 0, super.style, 0, 6)
        val parts =
            decomposedWith!!.getOrDefault(key)!!.split("%s".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        var argIndex = 0
        for (part in parts) {
            super.formattedText(part)
            if (argIndex < args.size) {
                argsIndices[argIndex] = super.components.size - 1
                component(args[argIndex])
                argIndex++
            }
        }
    }

    fun update(session: Session) {
        val currentLanguage = session.languageManager.language
        if (decomposedWith !== currentLanguage) {
            decomposedWith = currentLanguage
            decompose()
        }
    }

    fun getArg(index: Int): Component {
        return args[argsIndices[index]]
    }

    override val style: Style
        get() = Style.Companion.DEFAULT
}
